<?xml version="1.0"?>
<launch>

  <!--Robot description parameters -->
  <arg name="joint_limit_params"
       default="$(find ur_description)/config/ur5/joint_limits.yaml"/>
  <arg name="kinematics_params"
       default="$(find ur_description)/config/ur5/default_kinematics.yaml"/>
  <arg name="physical_params"
       default="$(find ur_description)/config/ur5/physical_parameters.yaml"/>
  <arg name="visual_params"
       default="$(find ur_description)/config/ur5/visual_parameters.yaml"/>
  <arg name="transmission_hw_interface"
       default="hardware_interface/PositionJointInterface"/>
  <arg name="safety_limits"	default="false"/>
  <arg name="safety_pos_margin"	default="0.15"/>
  <arg name="safety_k_position"	default="20"/>

  <param name="robot_description"
	 command="$(find xacro)/xacro '$(find ur_gazebo)/urdf/ur.xacro'
		  joint_limit_params:=$(arg joint_limit_params)
		  kinematics_params:=$(arg kinematics_params)
		  physical_params:=$(arg physical_params)
		  visual_params:=$(arg visual_params)
		  transmission_hw_interface:=$(arg transmission_hw_interface)
		  safety_limits:=$(arg safety_limits)
		  safety_pos_margin:=$(arg safety_pos_margin)
		  safety_k_position:=$(arg safety_k_position)"/>

  <!-- Robot state publisher -->
  <arg name="tf_prefix" default=""/>
  <arg name="tf_pub_rate" default="125"/>
  <node name="robot_state_publisher"
	pkg="robot_state_publisher" type="robot_state_publisher">
    <param name="publish_frequency" type="double" value="$(arg tf_pub_rate)" />
    <param name="tf_prefix"			  value="$(arg tf_prefix)" />
  </node>

  <!-- Start Gazebo and load the empty world if requested to do so -->
  <arg name="paused"	default="false"/>
  <arg name="gui"	default="true"/>
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name"	value="worlds/empty.world"/>
    <arg name="paused"		value="$(arg paused)"/>
    <arg name="gui"		value="$(arg gui)"/>
  </include>

  <!-- Spawn the model loaded earlier in the simulation just started -->
  <arg name="spawn_z" default="0.1"/>
  <node name="spawn_gazebo_model" pkg="gazebo_ros" type="spawn_model"
	args="-urdf
	      -param robot_description
	      -model robot
	      -z $(arg spawn_z)"
	output="screen" respawn="false" />

  <!-- Load and start the controllers listed in the 'controllers' arg. -->
  <arg name="controller_config_file"
       default="$(find ur_gazebo)/config/ur5_controllers.yaml"/>
  <arg name="controllers"
       default="joint_state_controller pos_joint_traj_controller"/>
  <rosparam file="$(arg controller_config_file)"/>
  <node name="ros_control_controller_spawner"
	pkg="controller_manager" type="spawner"
	args="$(arg controllers)" output="screen" respawn="false" />

  <!-- Load other controllers, but do not start them -->
  <arg name="stopped_controllers"
       default="joint_group_pos_controller"/>
  <node name="ros_control_stopped_spawner"
	pkg="controller_manager" type="spawner"
	args="--stopped $(arg stopped_controllers)"
	output="screen" respawn="false" />

  <!-- Start moveit planner -->
  <remap from="/scaled_pos_joint_traj_controller/follow_joint_trajectory"
	 to="/pos_joint_traj_controller/follow_joint_trajectory"/>
  <include file="$(find ur5_moveit_config)/launch/move_group.launch">
    <arg name="load_robot_description"	value="false"/>
  </include>

  <!-- Start controller manager -->
  <node name="rqt_controller_manager"
	pkg="rqt_controller_manager" type="rqt_controller_manager"/>
  
  <!-- Start Rviz -->
  <include file="$(find ur5_moveit_config)/launch/moveit_rviz.launch">
    <arg name="rviz_config"	value="$(find aist_controllers
				       )/launch/moveit_servo/ur5.rviz"/>
  </include>

</launch>
